### PR TITLE
[Tool] Add patches and scripts to compile StarRocks on Ubuntu

### PR DESCRIPTION
**Pull request by jiacheng-celonis**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [x] Tool

## Which issues of this PR fixes ：

Fixes building StarRocks on Ubuntu 22.04.

## Problem Summary(Required) ：

Currently, StarRocks developers often use the [starrocks-dev-env](https://hub.docker.com/r/starrocks/dev-env) docker container to build and run StarRocks. The docker container is based off CentOS. It has been a long-standing issue in the community that building StarRocks on Ubuntu or MacOS does not work out of the box. Building StarRocks in the docker container has several limitations:
- Using IDE such as CLion with docker container is hard. Although [remote development](https://www.jetbrains.com/help/clion/remote-projects-support.html) can be set up, but the IDE delay is high and several features do not work as in local development.
- Docker container is run as the root user. They produce files as root and often cause file permission issues when trying to access and modify the files outside of the container.
- Developers typically have tools and scripts in their local development. Docker environment makes reusing those tools difficult.

This PR includes a few patches and scripts to successfully build StarRocks (both FE, BE and starrocks-test) on Ubuntu 22.04.

The high-level steps are:
- Pull starrocks-dev-env docker image
- Copy the `/var/local/thirdparty` files to local (Ubuntu)
- Run `thirdparty/build-thirdparty-ubuntu.sh` to rebuild `libbrpc` and `libevents` natively on Ubuntu.
- Run `./build.sh` with the right flags.

Notable mentions on the patches:
- Set compilation option `STARROCKS_CXX_COMMON_FLAGS="-no-pie"` otherwise the linker complained about  `(libhdfs.a) `relocation R_X86_64_32 against .rodata.str1.1' can not be used when making a PIE object`
- Recompile `libevents` on Ubuntu, otherwise linker complained about `undefined reference: sysctl` on this [line](https://github.com/libevent/libevent/blob/8f47d8de281b877450474734594fdc0a60ee35d1/arc4random.c#L204).
- Recompile `libbrpc` with patches on Ubuntu, otherwise linker complained about `undefined reference to _dl_sym'`
  - The libbrpc patch is basically https://github.com/apache/incubator-brpc/pull/1783 https://github.com/apache/incubator-brpc/pull/1784. The libbrpc compilation issue has been fixed in the main, but StarRocks is still using an older version 0.9.7 so we need the patch.
- A few minor fixes in StarRocks code to make the code compile with g++-11 on Ubuntu.

The PR added a few additional scripts to assist the Ubuntu build:
- `vars-ubuntu.sh`
- `down-thirdparty-ubuntu.sh`
- `build-thirdparty-ubuntu.sh`

These scripts are modified from the existing scripts.

## How to build on Ubuntu?

1. Copy /var/local/thirdparty from starrocks-dev-env docker
	```
    cd thirdparty
	sudo ./copy_from_docker.sh
	sudo chown -R $USERNAME /var/local/thirdparty
	```

2. Rebuild libbrpc and libevents on Ubuntu
	```
	cd thirdparty
    JAVA_HOME=/usr/lib/jvm/java-11-openjdk-amd64  STARROCKS_GCC_HOME=/usr ./build-thirdparty-ubuntu.sh
	```
3. Build StarRocks BE
	```
	STARROCKS_CXX_COMMON_FLAGS="-no-pie" STARROCKS_GCC_HOME=/usr JAVA_HOME=/usr/lib/jvm/java-11-openjdk-amd64 STARROCKS_THIRDPARTY=/var/local/thirdparty ./build.sh --be
	```

4. Run StarRocks BE
	```
	LD_LIBRARY_PATH=/usr/lib/jvm/java-11-openjdk-amd64/lib/server output/be/start_be.sh
	```

For more details on how to build FE and how to set up CLion and IntelliJ on StarRocks on Ubuntu, please refer to this documentation: https://docs.google.com/document/d/1RzUweJ9NlN8pbuadwBFoROGmqhs3Z4ZCTh73TdQAH6k/

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Review by imay — DISMISSED** [review]

**Comment by chaoyli** [comment]
@jiacheng-celonis Thanks for your contribution. Could you sign the CLA?

File: thirdparty/vars-ubuntu.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+export TP_JAR_DIR=$TP_INSTALL_DIR/lib/jar
+
+#####################################################
+# Download url, filename and unpacked filename
+# of all thirdparties
+#####################################################
+
+# libevent
+# the last release version of libevent is 2.1.8, which was released on 26 Jan 2017, that is too old.
+# so we use the master version of libevent, which is downloaded on 22 Jun 2018, with commit 24236aed01798303745470e6c498bf606e88724a
```
Comment (chaoyli) [inline thirdparty/vars-ubuntu.sh:57]: I think we can upgrade the libevent and libbrpc to a higher version in StarRocks.

**Comment by decster** [comment]
> Set compilation option STARROCKS_CXX_COMMON_FLAGS="-no-pie" otherwise the linker complained about (libhdfs.a) relocation R_X86_64_32 against .rodata.str1.1' can not be used when making a PIE object`

Did not find this change in the patch, do you mean to add this manually, not in the build script? If it's required, we can add it to build script default.

Also I think we should change build script directly(looks like we just need to upgrade brpc, will make it works on both ubuntu&centos), and without introducing new scripts which rely on docker env, did you find some problem rebuilding all third-party libraries on ubuntu?

**Comment by jiacheng-celonis** [comment]
> > Set compilation option STARROCKS_CXX_COMMON_FLAGS="-no-pie" otherwise the linker complained about (libhdfs.a) relocation R_X86_64_32 against .rodata.str1.1' can not be used when making a PIE object`
> 
> Did not find this change in the patch, do you mean to add this manually, not in the build script? If it's required, we can add it to build script default.

Before, I was passing STARROCKS_CXX_COMMON_FLAGS="-no-pie" when I invoke `build.sh`: https://docs.google.com/document/d/1RzUweJ9NlN8pbuadwBFoROGmqhs3Z4ZCTh73TdQAH6k/edit#bookmark=id.y7swc2lnnud1

I wasn't sure if -no-pie is what StarRocks wants by default. Some older gcc versions make -no-pie binary by default, but on my gcc-11, it makes pie the default, so I have to explicitly set -no-pie.

I added "-no-pie" to `CMakeLists.txt` in [`51753ab` (#12358)](https://github.com/StarRocks/starrocks/pull/12358/commits/51753ab1bcdfb6565321e784a4e53d7dcf11bb4d), please check if that is okay.

> 
> Also I think we should change build script directly(looks like we just need to upgrade brpc, will make it works on both ubuntu&centos), and without introducing new scripts which rely on docker env, did you find some problem rebuilding all third-party libraries on ubuntu?

I think upgrading brpc will help. The patch for bprc is not necessary with their latest version.

I haven't tried rebuilding all third-party libraries on Ubuntu yet. I wanted to reuse the libraries from the dev container to save build time. But building all third-party libraries on Ubuntu will be a good exercise for another day.

**Comment by jiacheng-celonis** [comment]
> @jiacheng-celonis Thanks for your contribution. Could you sign the CLA?

Done

File: thirdparty/vars-ubuntu.sh (hunk 1 above)
Comment (jiacheng-celonis) [reply thirdparty/vars-ubuntu.sh:57]: I think upgrading libbrpc will certainly help. For libevent, I don't know if the `undefined reference: sysctl` error will be fixed by the upgrade.

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by chaoyli** [comment]
run starrocks_be_unittest

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by jiacheng-celonis** [comment]
The clang-tidy error seems like a false alert?

image

**Comment by chaoyli** [comment]
> The clang-tidy error seems like a false alert?
> 
> image

@jiacheng-celonis You can ignore the clang-tidy, there may be some mistakes to address. Just for a reference.

File: thirdparty/vars-ubuntu.sh (hunk 1 above)
Comment (chaoyli) [reply thirdparty/vars-ubuntu.sh:57]: I will upgrade the brpc to 1.3.0. Only 1.3.0 have the above two prs.

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Pull request closed without merging**